### PR TITLE
First-order finite-diff differentiation with float32 parameters

### LINKED ISSUE #1375: First-order finite-diff differentiation with float32 parameters results in floating-point errors
Currently, parameters of type `float32` will incur floating-point errors with finite-diff gradients.  The JAX interface defaults to `float32`, so the issue is most prominent there, but the problem could occur with any interface.

Inside `JacobianTape.numeric_pd`, the relevant parameter does get converted to `np.float64` when [the shift is added](https://github.com/PennyLaneAI/pennylane/blob/23b194ea61dd4158478ce06ef7f4719105f20b57/pennylane/tape/jacobian_tape.py#L265).  But for the calculation of [`"y0"`](https://github.com/PennyLaneAI/pennylane/blob/23b194ea61dd4158478ce06ef7f4719105f20b57/pennylane/tape/jacobian_tape.py#L536), the parameter values are not converted.

Because of this, finite differences will be accurate when `order=2`, but not when `order=1` (the default).

I propose fixing this by casting parameters to `np.float64` for calculation of `"y0"`:

```
                params_f64 = np.array(params, dtype=np.float64)
                options["y0"] = np.asarray(self.execute_device(params_f64, device))
```

This fix assumes every value in `params` is a single scalar. In other words, `params` is not a ragged array. Are we willing to make that assumption?

### PR DESCRIPTION
**Pull request by albi3ro**
Fixes #1375

Currently, `float32` parameters, like those default in the JAX interface, might experience floating-point errors with `diff_method='finite-diff'` and `order=1`.  `JacobianTape.numeric_pd` converts all parameters to `float64`, but the calculation of `"y0"` does not have such a conversion.

Downside:
We assume that the tape parameters are all scalers.  If matrix gate parameters, like with `qml.Hermitian` or `qml.QubitUnitary`, are not set to untrainable, then this fix will give errors.  We should have a better method to prevent trainable parameters from ever being non-scalers than just discouraging it.

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -351,6 +351,8 @@ random_mat2 = rng.standard_normal(3, requires_grad=False)
 
 <h3>Bug fixes</h3>
 
+* Fixes floating point errors with `diff_method="finite-diff"` and `order=1` when parameters are `float32`.
+
```
Comment (albi3ro) [inline .github/CHANGELOG.md:355]:
```suggestion
[(#1381)](https://github.com/PennyLaneAI/pennylane/pull/1381)

```

**Review by josh146 — CHANGES_REQUESTED** [review]
Thanks @albi3ro! PR looks good from my end. Only suggestion: I recommend adding a test that passes under this fix, but would have failed previously, just to ensure this remains fixed in the future.

> We should have a better method to prevent trainable parameters from ever being non-scalers than just discouraging it.

Agree - I think the correct approach here is `requires_grad=False` by default, as it requires users to deliberately turn on differentiation. However, this is a big change.

File: tests/tape/test_qnode.py [hunk 2]
```diff
@@ -956,3 +956,31 @@ def circuit(a):
         res = circuit(0.8, shots=2)
         assert len(res) == 2
         assert dev.shots == 3
+
+
+def test_finitediff_float32(tol):
```
Comment (albi3ro) [inline tests/tape/test_qnode.py:961]: I was unsure of where to put this test.  It's not actually an interface specific thing, just a float32 thing.

Comment (albi3ro) [reply tests/tape/test_qnode.py:961]: Also, I don't know why `StronglyEntanglingLayers` has a problem but simpler circuits don't.

Comment (josh146) [reply tests/tape/test_qnode.py:961]: > I was unsure of where to put this test. It's not actually an interface specific thing, just a float32 thing.

It looks fine here :) Alternatively, it could also be placed in `test_jacobian_tape` (does a file like that exist?).

> Also, I don't know why StronglyEntanglingLayers has a problem but simpler circuits don't.

Weird 🤔 

Thanks @albi3ro! So just to confirm, this test fails on master?

**Review by josh146 — APPROVED** [review]
Test looks great, thanks @albi3ro!

File: tests/tape/test_qnode.py (hunk 2 above)
Comment (albi3ro) [reply tests/tape/test_qnode.py:961]: test fails on master.  

I tried writing it in terms of tapes, but then you have to expand `StronglyEntanglingLayers` and update parameters and set them trainable, so I just went the simpler route of testing at QNode level.
